### ISSUE TITLE
Jenkinsfile still being corrupted with spacing

### ISSUE THREAD

**Issue by hazendaz** [openrewrite/rewrite-static-analysis#233]
I think this https://github.com/openrewrite/rewrite/pull/3789 needs to be merged so this stops happening as other items noted to fix the issue where Jenkinsfile is corrupted has not been resolved.

Using releases from today (1/9/2024).

Recipe ran

org.openrewrite.staticanalysis.CodeCleanup
org.openrewrite.java.format.NoWhitespaceBefore

This is different than the original condition as it hit in another area.  So likely the NoWhitespaceBefore is the new problem.  We have been stuck for months with this and inability to release.  We use open rewrite via pull requests on an automated pipeline running and corrupting the files doesn't help us try to make the process better.

Given this is a bit of whack a mule here, I think its best to revisit your original PR and simply don't touch Jenkinsfile until its understood better how to handle groovy.

My original issue was here 
- https://github.com/openrewrite/rewrite/issues/3783

versions used
- rewrite maven plugin 5.18.0
- static analysis 1.2.0

The source of the issue or recipe's ran are both in https://github.com/openrewrite/rewrite-static-analysis.

**Comment by timtebeek** [comment]
Hi @hazendaz ; Sorry to hear there continue to be issues with Jenkinsfiles for you! I didn't recall that you were still blocked by this issue, and a quick glance at board didn't show any immediate related issues: https://github.com/orgs/openrewrite/projects/4/views/36?filterQuery=label%3A%22parser-groovy%22
Would you be open to log the issue you're having, such that we can at least track that going forward?

Ideally we do want to parse Jenkinsfiles as Groovy, especially with an eye towards https://github.com/openrewrite/rewrite-jenkins/ . Did you already try setting a plaintext mask that includes Jenkinsfiles on the Maven or Gradle plugin configuration that you're using? That might provide some immediate relief, since the next full release might be a few weeks out.

Beyond that I'm open to for now indeed stop parsing Jenkinsfiles as Groovy, unless indeed the masks above help you further.
Any additional thoughts on this @knutwannheden ?
- https://github.com/openrewrite/rewrite/pull/3789

**Comment by knutwannheden** [comment]
We would need to invest more in the Groovy formatter to make sure it behaves correctly. Temporarily we could modify the `NoWhitespaceBefore` and `NoWhitespaceAfter` recipes which are part of `CodeCleanup` to only modify Java sources, as we have already done with some other formatting recipes.

**Comment by hazendaz** [comment]
@timtebeek 

I've added the following

```xml
<configuration>
    <plainTextMasks>
        <plainTextMask>**/Jenkinsfile</plainTextMask>
    </plainTextMasks>
</configuration>
```

That does do the trick and made it not write that file.  So I'm happy with that.  I'll go a little further here in case you want to keep the ticket open but you can close it too as this is good enough for us.

Do note I cannot share directly so most of this is just discussion or what I can show.

For history of what we are doing just for clarity and since I looked to check and I'm sure you might suggest better ways these days :)

We are using maven to drive the library usage in our case currently.  Its more of a alpha/beta type usage we have been trying since November 2021.  First it was a on demand via our super pom which contains some nearly 100 plugins.  Then we slowly added that to our jenkins pipeline to run on demand in a chained way using the safer and more frequently well working built in recipes where if one was good and applied it would move onto the next.  We more recently finally put this into our normal toolchain so it runs the same chained way.  Longer term we expect to split this out to a more jenkins friendly way and allow users to pass in recipe's they might want to run along with some canned ones.  Its because of the fact now its in our normal pipeline its a bit more concerning when it goes wrong especially the Jenkinsfile since that one is critical to actually work.

As with the initial issue I raised, the 'library something' is removing the middle space (will show more at end).  We don't much use much of anything in our Jenkinsfiles.  They simply load 2 or 3 libraries and call a function from one of them to run.  We actually have the Jenkinsfile in fact locked via a pull request process if we need updates so we only have one no matter the number of repos it uses.  So it really cannot get messed up or need any updates we haven't already figured out where important.  So for us, scanning the Jenkinsfile with open rewrite while cool would do really nothing in our case.  It would be cool to rather scan our real files to which we do have maven wrapped around them for some special needs but they have not been running this 'bot' yet.

So while the problem is still present the work around you have provided is really good enough for us forever as it pertains to Jenkinsfile.

I did however run this in debug in maven to get the output.  As we have these driven by profiles, the command in this case is 'mvn process-sources -Prewrite-cleanup'.  We provided no configuration to open rewrite outside of the work around you have provided to fix the issue.  We store the definition of open rewrite in dependency management and have 7 recipe's and an internal build tool jar as dependencies.  For the profile, it adds the two recipes: org.openrewrite.staticanalysis.CodeCleanup and org.openrewrite.java.RemoveUnusedImports.  The later we stuck on everything to get rid of 'star imports'.  Its execution is on the 'run' goal with process sources phase.

Configuration shows the two active recipes I noted.  There are no active styles.  Checkstyle detection is set to true, it is looking for a rewrite.yml although I don't have one and wasn't aware one existed so maybe that is a good thing to know..., no exclusions are listed, its not set to fail on invalid recipe's, (when not working, plain test masks was blank), rewrite skip is false, run per submodule is false, size threshold mb is 10, and skip maven parsing is false.

So a non working (ie not having the '<plainTextMask>**/Jenkinsfile</plainTextMask>') results in our file getting corrupted like this.

library'someLibrary@version'

or with longer form

libraryidentifier: 'repo@branch', changelog: false, retriever: modernSCM([rest of that])

Main point is that its the work 'library' that seems to trip this up.  I presume prior observations are correct its groovy's non requirement of ( where space is used instead.  Presumably we could have just added () around this and been ok but typically if you follow jenkins setup they show it done this way in every doc I've seen from them.  So really that would have been an option for us to fix as well although a bit more complicated to cycle that around.

So for grins I added () around the calls, it does not touch the file now in that case.  So I think the crux of this is that it just doesn't handle that type of groovy usage very well.  I did not test that specifically I could if needed but see no reason why Jenkins would dislike calling in more standard java way to methods.

**Comment by hazendaz** [comment]
I'll test the problematic condition against our real groovy files probably tomorrow it see if they have the same issues or not.  In such a case, I'll let you know but those I can just stop using spaces instead of () as only one copy exists so its not hard to just fix to get around any such issue.  And personally I prefer the java styling for consistency being that I'm a java developer first :)

**Comment by timtebeek** [comment]
Hi @hazendaz ; thanks for the detailed insight into how you're using the tools over there, and your efforts in helping fix this issue. Glad to hear the workaround for now at least works, but yes indeed there's an issue that we'd still like to solve.

I took a look just now and indeed it's painfully obvious: [the NoWhitespaceBefore recipe](https://docs.openrewrite.org/recipes/java/format/nowhitespacebefore) does exactly what it does on the tin for Groovy too, as Knut had already pointed out before. Should be fixed in
- https://github.com/openrewrite/rewrite/pull/3940

Hope that helps someone else, even if you're skipping Jenkinsfiles for now. :)
